Library/OcCpuLib/OcCpuLib.c

```diff
@@ -107,6 +88,161 @@ ScanThreadCount (
     }
   }
 
+  return Status;
+}
+
+STATIC
+EFI_STATUS
+ScanFrameworkMpServices (
+  IN  FRAMEWORK_EFI_MP_SERVICES_PROTOCOL  *FrameworkMpServices,
+  OUT OC_CPU_INFO                         *Cpu,
+  OUT UINTN                               *NumberOfProcessors,
+  OUT UINTN                               *NumberOfEnabledProcessors
+  )
+{
+  EFI_STATUS           Status;
+  UINTN                Index;
+  EFI_MP_PROC_CONTEXT  Context;
+  UINTN                ContextSize;
+
+  ASSERT (FrameworkMpServices != NULL);
+  ASSERT (Cpu != NULL);
+  ASSERT (NumberOfProcessors != NULL);
+  ASSERT (NumberOfEnabledProcessors != NULL);
+
+  ContextSize = sizeof (Context);
+
+  Status = FrameworkMpServices->GetGeneralMPInfo (
+    FrameworkMpServices,
+    NumberOfProcessors,
+    NULL,
+    NumberOfEnabledProcessors,
+    NULL,
+    NULL
+    );
+
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  if (*NumberOfProcessors == 0) {
+    return EFI_NOT_FOUND;
+  }
+
+  //
+  // This code assumes that all CPUs have same amount of cores and threads.
+  //
+  for (Index = 0; Index < *NumberOfProcessors; ++Index) {
+    Status = FrameworkMpServices->GetProcessorContext (
+      FrameworkMpServices,
+      Index,
+      &ContextSize,
+      &Context
+      );
+
+    if (EFI_ERROR (Status)) {
+      //
+      // It might make sense to error and exit here.
+      //
+      continue;
+    }
+
+    if (Context.PackageNumber + 1 >= Cpu->PackageCount) {
+      Cpu->PackageCount = (UINT16) (Context.PackageNumber + 1);
+    }
+
+    //
+    // According to the FrameworkMpServices header, EFI_MP_PROC_CONTEXT.NumberOfCores is the
+    // zero-indexed physical core number for the current processor. However, Apple appears to
+    // set this to the total number of physical cores (observed on Xserve3,1 with 2x Xeon X5550).
+    //
+    // This number may not be accurate; on MacPro5,1 with 2x Xeon X5690, NumberOfCores is 16 when
+    // it should be 12 (even though NumberOfProcessors is correct). Regardless, CoreCount and
+    // ThreadCount will be corrected in ScanIntelProcessor.
+    //
+    // We will follow Apple's implementation, as the FrameworkMpServices fallback was added for
```

Comment:
@Download-Fritz given that Mac Pro values are also trash, should we really do that? Perhaps avoid core/thread calculation code at all?

Reply:
Do we even need them? I remember seeing a bunch of values in the OcCpuLib struct that were never really used, maybe we should clean them up at some point and keep only those crucial to identify the hardware and/or debug detection required for macOS

Reply:
Yes, they should be in use by OcSmbiosLib.

Reply:
Leave it unchanged for now, I think we could always remove it later if it causes issues.

Reply:
Alright, will do.